mapboxgl/layers.py

```diff
@@ -0,0 +1,745 @@
+import json
+
+import numpy
+
+from mapboxgl.utils import color_map, height_map
+from mapboxgl import templates
+from mapboxgl.utils import img_encode, numeric_map
+
+
+class AbstractLayer():
+    """Abstract class for layers"""
+
+    def __init__(self,
+                 data,
+                 opacity=1,
+                 below_layer='',
+                 legend=True,
+                 legend_layout='vertical',
+                 legend_gradient=False,
+                 legend_text_numeric_precision=None,
+                 legend_key_shape='square',
+                 min_zoom=0,
+                 max_zoom=24,
+                 *args,
+                 **kwargs):
+        """Abstract class for layers
+
+                :param opacity: opacity of map data layer
+                :param below_layer: the layer to put this one below
+                :param legend: boolean for whether to show legend on map
+                :param legend_layout: determines if horizontal or vertical legend used
+                :param legend_gradient: boolean to determine if legend keys are discrete or gradient
+                :param legend_fill: string background color for legend, default is white
+                :param legend_header_fill: string background color for legend header (in vertical layout), default is #eee
+                :param legend_text_numeric_precision: decimal precision for numeric legend values
+                :param legend_key_shape: shape of the legend item keys, default varies by viz type; one of square, contiguous_bar, rounded-square, circle, line
+                """
+        self.data = data
+        self.opacity = opacity
+        self.below_layer = below_layer
+        self.legend = legend
+        self.legend_layout = legend_layout
+        self.legend_gradient = legend_gradient
+        self.legend_text_numeric_precision = legend_text_numeric_precision
+        self.legend_key_shape = legend_key_shape
+        self.min_zoom = min_zoom
+        self.max_zoom = max_zoom
+
+        self.layer_id = None
+        self.template = None
+        self.label_property = None
+
+    def create_html(self):
+        options = dict(
+            opacity=self.opacity,
+            belowLayer=self.below_layer,
```

Comment:
I guess this is done because the variables end up in HTMl/JS files where camelCase is quite common. But actually the variables don't enter into JS, since they are just used as placeholder in the templating, so I guess you are right! It will take some time changing all the variable names though.. :)